linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala

```diff
@@ -104,14 +106,27 @@ private[emitter] final class JSGen(val semantics: Semantics,
   }
 
   def genIsInstanceOf(expr: Tree, typeRef: TypeRef)(
+      implicit globalKnowledge: GlobalKnowledge, pos: Position): Tree = {
+    val useInstanceOf = typeRef match {
+      case ClassRef(className) =>
+        className != NumberClass && // the only non-Object super*class* of hijacked classes
+        !globalKnowledge.isInterface(className)
+      case ArrayTypeRef(_, _)  =>
+        false
+    }
+    genIsAsInstanceOf(expr, typeRef, test = true, useInstanceOf)
+  }
+
+  def genIsInstanceOf(expr: Tree, typeRef: TypeRef, isScalaClass: Boolean)(
       implicit pos: Position): Tree =
-    genIsAsInstanceOf(expr, typeRef, test = true)
+    genIsAsInstanceOf(expr, typeRef, test = true, useInstanceOf = isScalaClass)
```

Comment:
Hum, isn't this wrong in the case where `typeRef == ClassRef(NumberClass)`?

Reply:
Yes, it is, but it is never called for `NumberClass`. I have renamed the method and removed the argument to make it clearer. I have also added specific tests for `isInstanceOf[Number]`.